src/Remapping/distributed_remapping.jl

```diff
@@ -144,65 +219,482 @@ function Remapper(target_hcoords, target_zcoords, space)
-                hypsography. If `false` (the default), interpolation will be based on the
-                reference z coordinate, i.e. will correspond to constant model levels.
-                `NaN`s are returned for values that are below the surface.
+function set_interpolated_values_kernel!(
+    out,
+    (I1, I2)::NTuple{2},
+    local_horiz_indices,
+    fv,
+)
+
+    index = threadIdx().x + (blockIdx().x - 1) * blockDim().x
+    index > length(out) && return nothing
+
+    h = local_horiz_indices[index]
+    _, Nq = size(I1)
+
+    out[index] = 0
+    for j in 1:Nq, i in 1:Nq
+        out[index] +=
+            I1[index, i] * I2[index, j] * fv[i, j, nothing, nothing, h]
+    end
+
+    return nothing
+end
+
+function set_interpolated_values_kernel!(
+    out::AbstractArray,
+    (I,)::NTuple{1},
+    local_horiz_indices,
+    fv,
+)
+
+    index = threadIdx().x + (blockIdx().x - 1) * blockDim().x
+    index > length(out) && return nothing
+
+    h = local_horiz_indices[index]
+    _, Nq = size(I)
+
+    out[index] = 0
+    for i in 1:Nq
+        out[index] += I[index, i] * fv[i, nothing, nothing, nothing, h]
+    end
+
+    return nothing
+end
+
+"""
+    _apply_mpi_bitmask!(remapper::Remapper)
+
+Change to local (private) state of the `remapper` by applying the MPI bitmask and reconstructing
+the correct shape for the interpolated values.
+
+Internally, `remapper` performs interpolation on a flat list of points, this function moves points
+around according to MPI-ownership and the expected output shape.
+"""
+function _apply_mpi_bitmask!(remapper::Remapper)
+    if isnothing(remapper.target_zcoords)
+        # _interpolated_values[remapper.local_target_hcoords_bitmask] returns a view on
+        # space we want to write on
+        remapper._interpolated_values[remapper.local_target_hcoords_bitmask] .=
+            remapper._local_interpolated_values
+    else
+        # interpolated_values is an array of arrays properly ordered according to the bitmask
+
+        # _interpolated_values[remapper.local_target_hcoords_bitmask, :] returns a
+        # view on space we want to write on
+        remapper._interpolated_values[
+            remapper.local_target_hcoords_bitmask,
+            :,
+        ] .= remapper._local_interpolated_values
+    end
+end
+
+"""
+    _reset_interpolated_values!(remapper::Remapper)
+
+Reset the local (private) state in `remapper`. This function has to be called before performing
+interpolation.
+"""
+function _reset_interpolated_values!(remapper::Remapper)
+    fill!(remapper._interpolated_values, 0)
+end
+
+"""
+    _collect_interpolated_values!(remapper::Remapper)
+
+Perform an MPI call to aggregate the interpolated points from all the MPI processes and save
+the result in the local state of the `remapper`. Only the root process will return the
+interpolated data.
+"""
+function _collect_interpolated_values!(remapper::Remapper)
```

Comment:
Can we just manually inline this function? It's type-unstable, and I think the code will be clearer with it manually inlined.

Reply:
We discussed this a bit too. using `reduce!` would be type stable.

@Sbozzolo, since this is kind of internals, I'm fine with leaving this improvement for a later iteration.

Reply:
`collected_interpolated_values!` now uses `reduce!`